### PR TITLE
delete objects based on label

### LINKED ISSUE #255: Improving `down` to handle Volumes
In 7349dc9 i'm proposing adding --emptyvols to the up/down operations. This raises a bigger question on whether we should enable passing the other **convert** options so that they can be used with up/down - this can be useful, but maybe not something we want to do.

### PR DESCRIPTION
**Pull request by procrypt**
@kadel @containscafeine @surajssd @cdrage 

This PR uses the `SelectorFromSet()` function which takes a`label or selector` as an argument returns all the object that uses that `label or selector`. Once we get the object that uses a particular label we can further do the delete operation on them. This is similar to `kubectl delete <object> --selector=<key>=<value>`. Also the label has been modified from `service` to `io.kompose.service`.
Thanks @kadel for all the help :)
Review please.
Fixes #255

### CODE REVIEW

**Comment by cdrage** [comment]
Hi @procrypt 

Just two things before reviewing!

1. Please split the commits, one for updating vendoring, and the other the actual code you're modifying. Otherwise, it's really hard to review on GitHub / `gif dif`

2. Please update your Git commit messages outlining what's changed / more descriptive rather than `Fixes #255`. You can do this with `git commit --amend`

**Comment by procrypt** [comment]
Hey @cdrage I have split the commit into two different commit one for the _code_ and other for _vender_.
Also I have elaborated what this PR is doing.
Ready for review :)

**Comment by cdrage** [comment]
@procrypt 

Commit titles shouldn't have underscores in them, and it looks like you haven't put your updated message into your commit, you need to `git commit --amend`, add then `git push -f` it.

**Comment by procrypt** [comment]
@cdrage Done :D

**Comment by cdrage** [comment]
I wasn't excepting the labels to appear when using `kompose convert`, I've typed up my concerns in the issue #255 ! 

Great code however, I don't see any problems with it and I've tested the PR :+1:

**Comment by cdrage** [comment]
I think we should get in the habit of commenting on our code, mind adding a few lines @procrypt to your implementation?

File: pkg/transformer/kubernetes/kubernetes.go [hunk 1]
```diff
@@ -633,37 +635,67 @@ func (k *Kubernetes) Undeploy(komposeObject kobject.KomposeObject, opt kobject.C
 		switch t := v.(type) {
 		case *extensions.Deployment:
 			//delete deployment
-			rpDeployment, err := kubectl.ReaperFor(extensions.Kind("Deployment"), client)
+			label := labels.SelectorFromSet(labels.Set(map[string]string{"io.kompose.service": t.Name}))
```
Comment (kadel) [inline pkg/transformer/kubernetes/kubernetes.go]: `io.kompose.service` is used in a lot of places. maybe we could use `transformer.ConfigLabels` to handle all the labels.
Or other option would be to create new function or constant that will hold that string

**Comment by procrypt** [comment]
@kadel Updated PR.

**Comment by kadel** [comment]
this will require new vendor update :-(

I think that easiest thing to do,  will be remove 'update vendor' commit from your branch, do rebase, and than update vendor again ;-)

**Comment by procrypt** [comment]
@kadel Done all green now :D

File: pkg/transformer/kubernetes/kubernetes.go [hunk 2]
```diff
@@ -705,37 +708,67 @@ func (k *Kubernetes) Undeploy(komposeObject kobject.KomposeObject, opt kobject.C
 		switch t := v.(type) {
 		case *extensions.Deployment:
 			//delete deployment
-			rpDeployment, err := kubectl.ReaperFor(extensions.Kind("Deployment"), client)
+			label := labels.SelectorFromSet(labels.Set(map[string]string{selector: t.Name}))
```
Comment (kadel) [inline pkg/transformer/kubernetes/kubernetes.go]: This and next line is same for every object. You can safely move them outside of this `switch` and `for`, than you won't have to repeat same code multiple times 😉

File: pkg/transformer/openshift/openshift.go [hunk 3]
```diff
@@ -509,11 +514,20 @@ func (o *OpenShift) Undeploy(komposeObject kobject.KomposeObject, opt kobject.Co
 		switch t := v.(type) {
 		case *imageapi.ImageStream:
 			//delete imageStream
-			err = oclient.ImageStreams(namespace).Delete(t.Name)
+			label := labels.SelectorFromSet(labels.Set(map[string]string{selector: t.Name}))
```
Comment (kadel) [inline pkg/transformer/openshift/openshift.go]: same think as in kubernetes, you are repeating same lines

File: pkg/transformer/kubernetes/kubernetes.go [hunk 4]
```diff
@@ -705,37 +708,67 @@ func (k *Kubernetes) Undeploy(komposeObject kobject.KomposeObject, opt kobject.C
...
+			komposeLabel := map[string]string{selector: t.Name}
+			if reflect.DeepEqual(pvc.Items[0].Labels, komposeLabel) {
+				err = client.PersistentVolumeClaims(namespace).Delete(t.Name)
+				if err != nil {
+					return err
+				}
+				log.Infof("Successfully deleted PersistentVolumeClaim: %s", t.Name)
+			}
 
 		case *extensions.Ingress:
```
Comment (kadel) [inline pkg/transformer/kubernetes/kubernetes.go]: I think we should do same think for `Ingress` and `Pods` also.

File: pkg/transformer/openshift/openshift.go [hunk 5]
```diff
@@ -524,33 +538,60 @@ func (o *OpenShift) Undeploy(komposeObject kobject.KomposeObject, opt kobject.Co
...
+			komposeLabel := map[string]string{selector: t.Name}
+			if reflect.DeepEqual(pvc.Items[0].Labels, komposeLabel) {
+				err = kclient.PersistentVolumeClaims(namespace).Delete(t.Name)
+				if err != nil {
+					return err
+				}
+				log.Infof("Successfully deleted PersistentVolumeClaim: %s", t.Name)
+			}
 
 		case *routeapi.Route:
```
Comment (kadel) [inline pkg/transformer/openshift/openshift.go]: Don't forget about Pods and Routes ;-)

File: pkg/transformer/kubernetes/kubernetes.go [hunk 6]
```diff
@@ -695,6 +697,7 @@ func (k *Kubernetes) Undeploy(komposeObject kobject.KomposeObject, opt kobject.C
 	if err != nil {
 		return errors.Wrap(err, "k.Transform failed")
 	}
+	selector := "io.kompose.service"
```
Comment (cdrage) [inline pkg/transformer/kubernetes/kubernetes.go]: not to nitpick, but should this instead be a constant and at the top of the file? (outside the scope) ping @kadel

Comment (kadel) [inline pkg/transformer/kubernetes/kubernetes.go]: @cdrage you are right. I was thinking about it, but forgot to commend on it :-)

This should be even somewhere else (maybe const in transformer pkg?), so [ConfigLabels](https://github.com/kubernetes-incubator/kompose/blob/5cb598fa5bed08ffe52b4a02fd299fbf3de373e3/pkg/transformer/utils.go#L118) function can use it, and we don't have same magic string in two places.

File: pkg/transformer/kubernetes/kubernetes.go (hunk 2 above)
Comment (procrypt) [inline pkg/transformer/kubernetes/kubernetes.go]: @kadel If we put
 ```golang 
label := labels.SelectorFromSet(labels.Set(map[string]string{selector: t.Name}))
```
outside `switch` and `for` how do we access the name of the object we want to delete created, here we are accessing it using `t.Name`.

Comment (kadel) [inline pkg/transformer/kubernetes/kubernetes.go]: Than you can do it like this:
Keep it in for, but move it outside switch.

You can access object metadata fields by casting it to `meta.Object`.
```go
# import "k8s.io/kubernetes/pkg/api/meta"
v.(meta.Object).GetName()
```

**Comment by kadel** [comment]
if nothing is deployed and  call `kompose down` is called:
```
./kompose down
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/kubernetes-incubator/kompose/pkg/transformer/kubernetes.(*Kubernetes).Undeploy(0xc4202cc180, 0xc42033e8a0, 0x1c625c5, 0x7, 0x100, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/tomas/dev/go/src/github.com/kubernetes-incubator/kompose/pkg/transformer/kubernetes/kubernetes.go:740 +0x14d6
github.com/kubernetes-incubator/kompose/pkg/app.Down(0x100, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1, 0xc42025fe50, 0x1, 0x1, ...)
	/home/tomas/dev/go/src/github.com/kubernetes-incubator/kompose/pkg/app/app.go:288 +0x2be
github.com/kubernetes-incubator/kompose/cmd.glob..func5(0x2a3b200, 0x2a69368, 0x0, 0x0)
	/home/tomas/dev/go/src/github.com/kubernetes-incubator/kompose/cmd/down.go:52 +0x4f
github.com/kubernetes-incubator/kompose/vendor/github.com/spf13/cobra.(*Command).execute(0x2a3b200, 0x2a69368, 0x0, 0x0, 0x2a3b200, 0x2a69368)
	/home/tomas/dev/go/src/github.com/kubernetes-incubator/kompose/vendor/github.com/spf13/cobra/command.go:603 +0x22b
github.com/kubernetes-incubator/kompose/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x2a3b420, 0xc420070058, 0x0, 0xc4204b9f48)
	/home/tomas/dev/go/src/github.com/kubernetes-incubator/kompose/vendor/github.com/spf13/cobra/command.go:689 +0x339
github.com/kubernetes-incubator/kompose/vendor/github.com/spf13/cobra.(*Command).Execute(0x2a3b420, 0x1, 0x1)
	/home/tomas/dev/go/src/github.com/kubernetes-incubator/kompose/vendor/github.com/spf13/cobra/command.go:648 +0x2b
github.com/kubernetes-incubator/kompose/cmd.Execute()
	/home/tomas/dev/go/src/github.com/kubernetes-incubator/kompose/cmd/root.go:92 +0x31
main.main()
	/home/tomas/dev/go/src/github.com/kubernetes-incubator/kompose/main.go:22 +0x20
```

**Comment by procrypt** [comment]
@kadel Ah this is bad, I guess I know the what is the problem, when we call `kompose down` the 
```golang
client.Deployments(namespace).List()            returns  *extensions.DeploymentList
client.Services(namespace).List()               returns  *api.ServiceList
client.PersistentVolumeClaims(namespace).List() returns  *api.PersistentVolumeClaimList
```
and the struct 
```golang
DeploymentList            contains  Items []Deployment
ServiceList               contains  Items []Service
PersistentVolumeClaimList contains  Items []PersistentVolumeClaim
```
and to get the `Label` we are accessing the first index of the each of slice of 
```golang
Deployment              using  deployment.Items[0].Labels
Service                 using  svc.Items[0].Labels
PersistentVolumeClaim   using  pvc.Items[0].Labels
```
So this works when we do `kompose up` and then do `kompose down` because the application has been already deployed and `kompose down` get what needed to find the `Label`, but if we do `kompose down` and there is no application deployed the there is are `objects`, `kompose` tries to access the first element of the 
```golang
DeploymentList 
ServiceList
PersistentVolumeClaimList
```
and since there is nothing like
```golang
deployment.Items[0].Labels
svc.Items[0].Labels
pvc.Items[0].Labels
```
It panics with `runtime error: index out of range`

**Comment by procrypt** [comment]
@kadel Fixed `kompose down`
```bash
$ kompose down
FATA[0000] Error while deleting application: service `mlbparks` not found
```

**Comment by procrypt** [comment]
@kadel @cdrage Updated the PR Please Review

File: pkg/transformer/kubernetes/kubernetes.go [hunk 7]
```diff
@@ -702,40 +705,78 @@ func (k *Kubernetes) Undeploy(komposeObject kobject.KomposeObject, opt kobject.C
...
-			rpDeployment, err := kubectl.ReaperFor(extensions.Kind("Deployment"), client)
+			deployment, err := client.Deployments(namespace).List(options)
 			if err != nil {
 				return err
 			}
-			//FIXME: gracePeriod is nil
-			err = rpDeployment.Stop(namespace, t.Name, TIMEOUT*time.Second, nil)
-			if err != nil {
-				return err
+			komposeLabel := map[string]string{transformer.Selector: t.Name}
```
Comment (kadel) [inline pkg/transformer/kubernetes/kubernetes.go]: You can move this outside of this switch. That you wont have to repeat same line in every case.

File: pkg/transformer/kubernetes/kubernetes.go [hunk 8]
```diff
@@ -702,40 +705,78 @@ func (k *Kubernetes) Undeploy(komposeObject kobject.KomposeObject, opt kobject.C
...
+			if len(deployment.Items) == 0 {
```
Comment (kadel) [inline pkg/transformer/kubernetes/kubernetes.go]: You can completely remove this this now.

Comment (procrypt) [inline pkg/transformer/kubernetes/kubernetes.go]: @kadel If we remove this line, there will be no output of `kompose down` is there is nothing deployed and you mentioned earlier that we want at least want to give an error in this situation.

Comment (kadel) [inline pkg/transformer/kubernetes/kubernetes.go]: That is OK. It is not an error if nothing is found.

Comment (procrypt) [inline pkg/transformer/kubernetes/kubernetes.go]: But we should warn the user about it.

**Comment by procrypt** [comment]
@kadel Updated and removed the check :wink:

**Pull request merged**